### PR TITLE
Adding base xcmp handler pallet

### PR DESCRIPTION
**Pull request by rhuttman**

### CODE REVIEW

File: pallets/xcmp-handler/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+		pub native: bool,
+		pub fee_per_second: u128,
+		pub instruction_weight: u64,
+	}
+
+	/// Stores XCM data for a chain/currency pair.
+	#[pallet::storage]
+	#[pallet::getter(fn get_xcm_chain_data)]
+	pub type XcmChainCurrencyData<T: Config> =
+		StorageDoubleMap<_, Twox64Concat, u32, Twox64Concat, T::CurrencyId, XcmCurrencyData>;
```
Comment (nvengal) [inline pallets/xcmp-handler/src/lib.rs:107]: Could you add a type alias for the `u32`? Maybe `ParaId` (unless that name is already overloaded)? It's nice to be explicit about what's being stored.

File: pallets/xcmp-handler/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+		#[pallet::weight(T::WeightInfo::remove_chain_currency_data())]
+		pub fn remove_chain_currency_data(
+			origin: OriginFor<T>,
+			para_id: u32,
+			currency_id: T::CurrencyId,
+		) -> DispatchResultWithPostInfo {
+			ensure_root(origin)?;
+
+			XcmChainCurrencyData::<T>::remove(para_id, currency_id);
+			Self::deposit_event(Event::XcmDataRemoved { para_id });
```
Comment (justinzhou93) [inline pallets/xcmp-handler/src/lib.rs:142]: also with currency ID

**Review by justinzhou93 — CHANGES_REQUESTED** [review]

File: pallets/xcmp-handler/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+		///Add or update XCM data for a chain/currency pair.
+		///For now we only support our native currency.
+		#[pallet::weight(T::WeightInfo::add_chain_currency_data())]
+		pub fn add_chain_currency_data(
+			origin: OriginFor<T>,
+			para_id: u32,
+			currency_id: T::CurrencyId,
+			xcm_data: XcmCurrencyData,
+		) -> DispatchResultWithPostInfo {
+			ensure_root(origin)?;
```
[the remark this answers is not in the document]
Comment (rhuttman) [reply pallets/xcmp-handler/src/lib.rs:120]: i _think_  that governance can act as sudo. We need to test that though. It is on my list,

File: pallets/xcmp-handler/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+mod weights;
+pub use weights::WeightInfo;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use codec::Decode;
+	use frame_support::{
+		dispatch::DispatchResultWithPostInfo,
+		pallet_prelude::{StorageDoubleMap, *},
```
[the remark this answers is not in the document]
Comment (rhuttman) [reply pallets/xcmp-handler/src/lib.rs:49]: huh I did not add that explicitly... thanks VS code for being silly

**Review by justinzhou93 — APPROVED** [review]
lgtm

File: pallets/xcmp-handler/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+			currency_id: T::CurrencyId,
+		) -> DispatchResultWithPostInfo {
+			ensure_root(origin)?;
+
+			if let Some(_) = XcmChainCurrencyData::<T>::get(para_id, currency_id) {
+				XcmChainCurrencyData::<T>::remove(para_id, currency_id);
+				Self::deposit_event(Event::XcmDataRemoved { para_id, currency_id });
+			} else {
+				Err(Error::<T>::CurrencyChainComboNotFound)?
+			}
```
Comment (nvengal) [inline pallets/xcmp-handler/src/lib.rs:155]: Not sure if syntactic sugar or magically saves a read. Probably just syntactic sugar.
```suggestion
			XcmChainCurrencyData::<T>::take(para_id, currency_id)
				.ok_or(Error::<T>::CurrencyChainComboNotFound)?;
			Self::deposit_event(Event::XcmDataRemoved { para_id, currency_id });
```

File: pallets/xcmp-handler/src/tests.rs [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+//*****************
+
+// add_chain_currency_data
+#[test]
+fn can_add_new_data() {
+	new_test_ext().execute_with(|| {
+		let currency_id = CurrencyId::Native;
+		let para_id: u32 = 1000;
+
+		if let Some(_) = XcmChainCurrencyData::<Test>::get(para_id, currency_id) {
```
Comment (nvengal) [inline pallets/xcmp-handler/src/tests.rs:32]:
```suggestion
		if XcmChainCurrencyData::<Test>::get(para_id, currency_id).is_some() {
```

**Review by nvengal — APPROVED** [review]
Added some non-blocking syntax suggestions

**Review by nvengal — APPROVED** [review]

**Pull request merged**